src/builder/ruleset_builder.cpp

```diff
@@ -0,0 +1,198 @@
+                final_base_rules_.emplace(builder->build(*actions_));
+            }
+        }
+    }
+
+    if ((current_changes & custom_rule_update) != change_set::none) {
+        final_user_rules_.clear();
+        // Initially, new rules are generated from their spec
+        for (const auto &[id, spec] : global_config.user_rules) {
+            rule_builder builder{id, spec};
+            if (builder.is_enabled()) {
+                final_user_rules_.emplace(builder.build(*actions_));
+            }
+        }
+    }
+
+    // Generate exclusion filters targetting all final rules
+    if ((current_changes & filters_update) != change_set::none) {
+        rule_filters_.clear();
+        input_filters_.clear();
+
+        // First generate rule filters
+        for (const auto &[id, filter] : global_config.rule_filters) {
+            auto rule_targets = resolve_references(filter.targets, final_base_rules_);
+            rule_targets.merge(resolve_references(filter.targets, final_user_rules_));
+
+            auto filter_ptr = std::make_shared<exclusion::rule_filter>(
+                id, filter.expr, std::move(rule_targets), filter.on_match, filter.custom_action);
+            rule_filters_.emplace(id, filter_ptr);
+        }
+
+        // Finally input filters
+        for (const auto &[id, filter] : global_config.input_filters) {
+            auto rule_targets = resolve_references(filter.targets, final_base_rules_);
+            rule_targets.merge(resolve_references(filter.targets, final_user_rules_));
+
+            auto filter_ptr = std::make_shared<exclusion::input_filter>(
+                id, filter.expr, std::move(rule_targets), filter.filter);
+            input_filters_.emplace(id, filter_ptr);
+        }
+    }
+
+    // Generate new processors
+    if ((current_changes & processors_update) != change_set::none) {
+        preprocessors_.clear();
+        postprocessors_.clear();
+
+        for (const auto &[id, spec] : global_config.processors) {
+            auto proc = processor_builder::build(id, spec, global_config.scanners);
+            if (spec.evaluate) {
+                preprocessors_.emplace(proc->get_id(), std::move(proc));
+            } else {
+                postprocessors_.emplace(proc->get_id(), std::move(proc));
+            }
+        }
+    }
+
+    if ((current_changes & change_set::rule_data) != change_set::none) {
+        rule_matchers_.clear();
+        for (const auto &[id, spec] : global_config.rule_data) {
+            rule_matchers_.emplace(id, matcher_builder::build(spec));
+        }
+    }
+
+    if ((current_changes & change_set::exclusion_data) != change_set::none) {
+        exclusion_matchers_.clear();
+        for (const auto &[id, spec] : global_config.exclusion_data) {
+            exclusion_matchers_.emplace(id, matcher_builder::build(spec));
+        }
+    }
+
+    auto rs = std::make_shared<ruleset>();
+    rs->insert_rules(final_base_rules_.items(), final_user_rules_.items());
```

Comment:
why are rules individually managed through `shared_ptr` instead of `final_base_rules_` being themselves behind a shared pointer (like the actions). Because AFAICT, whenever we have a `base_rule_update`, we rebuild the rules from scratch:

```cpp
    if ((current_changes & base_rule_update) != change_set::none) {
        final_base_rules_.clear();
        // ...
        for (const auto &builder : rule_builders) {
            if (builder->is_enabled()) {
                final_base_rules_.emplace(builder->build(*actions_));
            }
        }
```

Same applies to the user rules.

As mentioned earlier, I would feel much safer if these rules had their immutability enforced by the compiler, to prevent races.

Reply:
I'll address this in the next PR.